### PR TITLE
fixes #8857 - refactor Debian boot_files_uri into parent class to match CoreOS

### PR DESCRIPTION
**Pull request by johscheuer**
Refactor Debian boot_files_uri into parent class to match CoreOS. See Issue http://projects.theforeman.org/issues/8857

### CODE REVIEW

**Comment by domcleal** [comment]
[test]

**Comment by stbenjam** [comment]
[test]

File: test/unit/debin_test.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
+require 'test_helper'
+
+class DebianTest < ActiveSupport::TestCase
```
Comment (dLobatog) [inline test/unit/debin_test.rb]: The name of this file has a typo, "debin" instead of "debian". When you correct it, please put the test in `test/unit/operatingsystems/debian_test.rb`, just to keep a consistent hierarchy with the one in app/models

File: test/unit/ubuntu_test.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
+require 'test_helper'
+
+class UbuntuTest < ActiveSupport::TestCase
```
Comment (dLobatog) [inline test/unit/ubuntu_test.rb]: Same as for the Debian test, please move them (`coreos_test, debian_test, suse_test`) to a better suited subfolder `test/unit/operatingsystems/`

File: test/fixtures/operatingsystems.yml [hunk 3]
```diff
@@ -64,4 +64,24 @@ coreos:
   release_name: stable
   architectures: x86_64
   media: coreos
-  title: 'CoreOS 494.5.0'
\ No newline at end of file
+  title: 'CoreOS 494.5.0'
+
+debian7_0:
```
Comment (dLobatog) [inline test/fixtures/operatingsystems.yml]: Prefer adding new Factories or using existing ones, we are trying to reduce fixtures usage as much as possible.

File: test/unit/ubuntu_test.rb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    host.arch.operatingsystems << host.os
+
+    prefix = host.os.pxe_prefix(host.arch).to_sym
+    pxe_files = host.os.pxe_files(host.medium, host.arch)
+
+    kernel = { prefix => "http://sg.archive.ubuntu.com/dists/utopic/main/installer-amd64/current/images/netboot/ubuntu-installer/amd64/linux"  }
+    initrd = { prefix => "http://sg.archive.ubuntu.com/dists/utopic/main/installer-amd64/current/images/netboot/ubuntu-installer/amd64/initrd.gz" }
+    assert pxe_files.include?(kernel)
+    assert pxe_files.include?(initrd)
+  end
```
Comment (dLobatog) [inline test/unit/ubuntu_test.rb]: These tests are identical to the ones in `coreos_test.rb` and `debian_test.rb`. Could you create some sort of template for these and just pass as arguments the things that change? It looks like the only difference between these tests is the word "ubuntu" and the kernel, initrd urls.

File: test/unit/ubuntu_test.rb [hunk 5]
```diff
@@ -0,0 +1,45 @@
+require 'test_helper'
+
+class UbuntuTest < ActiveSupport::TestCase
+
+  test "os label for ubuntu" do
+    assert_equal "Ubuntu 14.10", operatingsystems(:ubuntu14_10).to_label
+  end
+
+  test "kernel location for 64bit arch ubuntu" do
+    host = FactoryGirl.create(:host, :operatingsystem => operatingsystems(:ubuntu14_10),
```
Comment (dLobatog) [inline test/unit/ubuntu_test.rb]: These three `host` creation calls can be just `.build`, which won't hit the database and make your tests faster. Same for `debian_test` (and `coreos_test`, although I understand that's not part of this commit). 

Also you could put these methods in a context with a setup block, like

``` ruby
context "some name" do 
  setup do 
   host = FactoryGirl.build(:host, :operatingsystem => operatingsystems(:ubuntu14_10),
     :architecture => architectures(:x86_64))
  end

  test ...
end
```

File: app/models/operatingsystems/debian.rb [hunk 6]
```diff
@@ -11,16 +15,7 @@ def preseed_path(host)
   end
 
   def boot_files_uri(medium, architecture, host = nil)
-    raise ::Foreman::Exception.new(N_("invalid medium for %s"), to_s) unless media.include?(medium)
```
Comment (dLobatog) [inline app/models/operatingsystems/debian.rb:14]: Nice! Just one thing, could you make the parent class raise a ::Foreman::Exception too? It seems like they were **only** raised when provided an invalid medium/architecture for Debian.

**Comment by dLobatog** [comment]
@johscheuer Nice job! There are only a few issues with the way tests are structured in my opinion, it works well although I have not thoroughly tested it yet. See the comments above and I'll test it a bit more, but it looks like :+1: from my side. :smile:

**Comment by johscheuer** [comment]
@eLobato Thanks :+1: :) I will have a look and fix these things.

**Comment by mmoll** [comment]
[test]

File: test/unit/operatingsystem_test.rb [hunk 7]
```diff
@@ -225,9 +225,9 @@ class OperatingsystemTest < ActiveSupport::TestCase
   end
 
   test "should find os name using free text search only" do
-    operatingsystems = Operatingsystem.search_for('OpenSuse')
+    operatingsystems = Operatingsystem.search_for('centos')
     assert_equal 1, operatingsystems.count
-    assert_equal operatingsystems(:suse), operatingsystems.first
+    assert_equal operatingsystems(:centos5_3), operatingsystems.first
   end
```
Comment (johscheuer) [inline test/unit/operatingsystem_test.rb:231]: I changed the os from Suse to centos, so we can safely remove the fixture for Suse

**Comment by mmoll** [comment]
[test]

**Comment by dLobatog** [comment]
[test]

**Comment by dLobatog** [comment]
Looks pretty good, I'll comment a few nitpicks I'll fix on merge just to leave some footprint. Just a sec...

File: app/models/operatingsystem.rb [hunk 8]
```diff
@@ -252,7 +252,7 @@ def downcase_release_name
   end
 
   def boot_files_uri(medium, architecture, host = nil)
-    raise (_("invalid medium for %s") % to_s) unless media.include?(medium)
+    raise ::Foreman::Exception.new(N_("invalid medium for %s"), to_s) unless media.include?(medium)
```
Comment (dLobatog) [inline app/models/operatingsystem.rb:255]: Change `to_s` to `self`. A rubocop cop we should use forces this, it's much more readable.

File: test/unit/operatingsystems/operatingsystems_test.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
+require 'test_helper'
+
+class OperatingsystemTest < ActiveSupport::TestCase
+  {:coreos => { 'os' => :coreos, 'arch' => :x86_64, 'expected' => 'CoreOS 494.5.0'},
```
Comment (dLobatog) [inline test/unit/operatingsystems/operatingsystems_test.rb:4]: Indentation (especially of these hashes) is a bit off, to be fixed.

**Comment by dLobatog** [comment]
Merged as 03363cd527a9f5d6fcada0b33969a2d5cede477b, fixes in e7605129b329264b8265fb5aaee942bf95aabbed, thanks @johscheuer!
